### ISSUE TITLE
stabilization effort

### ISSUE THREAD

**Issue by tomberek** [NixOS/nix#7701]
This is a DRAFT
Tracking for the stabilizing effort:

Under consideration when stabilizing
1) consider  with legacy CLI
2) review correctness of command
- [ ] do all `--help` options make sense

# Ready to go
- [ ] `nix store gc`

- [ ] `nix store ping`
- [ ] `nix store add-file`
- [ ] `nix store repair`

# Tasks
Update mapping to provide migration and usage guidance.

[CLI Guidelines](https://nixos.org/manual/nix/stable/contributing/cli-guideline.html)
[legacy CLI mapping](https://github.com/NixOS/nix/issues/4429)

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nix/pull/7740 / https://github.com/NixOS/nix/issues/5638 being hard I think it something to think about prior to command stabilization.

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting 2023-02-10:

Decision: work through one complete command at a time and write down the process. once finished and everyone is confident to reproduce that process on one's own, do the rest of the work in parallel.

<details><summary>Complete discussion</summary>

- @tomberek will clarify the exact things to do
   - @Ericson2314: the idea was to take one command at a time to figure out these things to begin with
   - @fricklerhandwerk: why don't we do it together for one command?
   - @Ericson2314: we already started with `nix store gc`
   - @fricklerhandwerk: let's take one hour together in the work meetings to attack this and write down what we do. when one command is finished (it is likely to take months at that pace), and everyone is confident to reproduce the process on one's own given the documentation we write, we can parallelise it for all the other commands
       - agreement
       - @roberth: several months are a blip in relation to how long it took to develop the new CLI anyway
   - @thufschmitt: will make a PR to put on the agenda for next time
       - won't be there myself though

</details>

**Comment by Ericson2314** [comment]
Two topics (with my PR fixes, if we like my solution) that I think would be good to track:

- Gating experimental settings on experimental features --- https://github.com/NixOS/nix/pull/7609
- Hiding freeform prose on experimental topics in `--help` ---  https://github.com/NixOS/nix/pull/7611

(My middle PR, https://github.com/NixOS/nix/pull/7610, corresponds to https://github.com/NixOS/nix/issues/7820 already listed here, so there is no need to link the issue directly.)

**Comment by roberth** [comment]
Meeting notes 2023-09-01


#### `nix store gc`

What about the gc-related queries?

- Maybe add `--dry-run` and `--json` to `nix store gc`?
- or to `nix store delete`
- How do we structure the "reverse" queries? Just gc roots? or more?
- Returning only *some* things can be confusing
- Returning everything may be slow
- @roberth: If we return to little, then the users has to backtrack with docs, and then still run the slow thing, worst of both worlds

#### `nix store ping`

What does `--json` return?
Why does `trusted` return an int? Change to bool-or-`null`

TODO copy note from comment on issue

Rename to `nix store info`.


#### `nix store add-file`

`add-path`?

@edolstra: benefit of explicit `add-file` is that it matches `fetchurl`
@ericson2314: a single command with multiple modes, flags seems more sensible
@thufschmitt: what's the use case?
@edolstra: generally prefetching
@roberth: also RFC 92 dynamic derivations?
@thufschmitt: seem similar
@ericsion2314: the varying hashing methods have always been a source of confusion to me. Making this explicit is helpful for learning
@edolstra: names?
@roberth: match the code?
Too long and inconvenient
@roberth: short flags that are neatly categorized in the docs?
@edolstra: should match derivation attr `recursive`, `flat`
@roberth: `recursive` should be the default?
Make it match the language? Then it should be `recursive` or `flat`
@roberth: ~`recursive-or-flat` as a mode?~ ~It names the language behavior~ wrong! The language uses `recursive` in its string coercions.
Should it match [`outputHashAlgo` / `outputHashMethod`](https://nixos.org/manual/nix/stable/language/advanced-attributes.html?highlight=outputHashAlgo#adv-attr-outputHash)?
@thufschmitt: It would be great for the language and CLI to agree
@ericson2314: I think it is very good if the CLI is "subliminally" teaching people the most compressed/efficient mental model by never missing and opportunity to connect things that are the same. (The less users need to consult the docs the better.) So if recursive hashing is deeply related to NAR hashing, they should share a name. (I took a lot of time to "academically" understand Nix and refine my mental model, but most users won't take the time to "relearn" that way. So we want to give them the right mental model from the beginning.)

@thufschmitt: Do we want key-value options or flags? (`--method=foo` or `--foo`)
@edolstra: Prefer brief options. Would like to avoid string-value for enums.
@roberth: This command is for scripts and power users, so brevity is not a priority.
@tomberek: What should be the CLI guideline?
@edolstra: enum key-value options can always be turned into single flags
@ericson2314: such flags are naturally booleans, whereas we're dealing with an enum
@roberth: key-value matches the language, matches derivations

@edolstra: We have precendent for flattening enum-valued options to mutually exclusive flags, e.g. sandbox
@roberth: for tab completion we want a consistent prefix
@edolstra: Implementation effort may be higher for enum case
@roberth: Should be close: either you specify more flags, or you specify a string mapping
@ericson2314: Would be happy to implement infra for enum options if still needed.
@ericson3214: Support `text` for manually adding derivations in aterm format (e.g. for RFC 92, or just for complete coverage)

@tomberek: Let's also update / clarify the CLI guidelines.

**Comment by anka-213** [comment]
> * Maybe add `--dry-run` and `--json` to `nix store gc`?

I would argue that almost all of the commands in the new CLI need a `--dry-run` flag (#7227)

**Comment by roberth** [comment]
Just had a conversation on mastodon, where `nix store gc` was compared to the legacy command with `-d`.
The lack of generation deletion is somewhat of a regression from the user perspective, so we can only really claim parity when we have a `nix profile` subcommand that does root deletion *and* gc.
Would `nix profile collect` be a good name? My reasoning is that it doesn't collect just the garbage.

**Comment by fricklerhandwerk** [comment]
`nix profile clean[up]`

What is it supposed to collect?

**Comment by thufschmitt** [comment]
@roberth : I think it would make sense to have both (a `nix store gc` that doesn't care about the profiles and `nix profile collect/cleanup` that also removes the old profiles (maybe in a more targeted way building on top of #7239?).

**Comment by Poly2it** [comment]
Why should `nix run --file` imply impure? Is it any more impure than passing an `--expr`? As I see it, it has been provided explicitly as an input by the user.

**Comment by sdht0** [comment]
What's the latest news on this effort? The last update here seems to be from October.

**Comment by edolstra** [comment]
Added replacements for `nix-store --export/--import` in https://github.com/NixOS/nix/pull/13969.

**Comment by x10an14** [comment]
How does (if it does at all) this issue address the `nix-build --attrs` flag which I can't (perhaps in error) find an equivalent for in `man nix3-build`?

```
$ nix run nixpkgs#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.16.9, NixOS, 25.11 (Xantusia), 25.11.20251002.7df7ff7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.5`
 - channels(root): `"nixos"`
 - channels(x10an14): `""`
 - nixpkgs: `/nix/store/9v6qa656sq3xc58vkxslqy646p0ajj61-source`
```
